### ISSUE TITLE
Need help converting code to callback function or async/await

### ISSUE THREAD

**Issue by Syntthetix** [nodejs/help#2891]
* **Node.js Version**: 12.16.1
* **OS**: Windows 10 Home
* **Scope (install, code, runtime, meta, other?)**: Code
* **Module (and version) (if relevant)**: N/A

Code:
(This function simply needs to be called in another file to retrieve an ID from an SQL database (the SQL part works perfectly fine, nothing to worry about there) and return the ID for use in that other file.)
```
let getChannel = (chanType) => {
  var chanID; // Value that needs to be assigned and returned to var that called the function in another file

  db.getConnection((err, con) => {
    con.query(`SELECT * FROM channels WHERE type = '${chanType}'`, (err, rows) => {
      if (err) throw err;

      chanID = rows[0].id;

      // Still holds value here
      con.release();
    });
  });
  return chanID; // reverts to undefined here
}
```
The code that I have right now has the problem of always returning `chanID` as `undefined`. From some research I did, I found that it has to do with the asynchronous nature of node.js, and that concept makes sense. The part of the code that actually reassigns the value of `chanID` is being executed too late because of all the asynchronous stuff, that much is simple enough.

Basically, I need to convert this code into a setup with either a callback or using async/await. I'm fairly new to Javascript and node.js though, so I'm having a lot of trouble figuring out how to actually do that. Every example I've found on how to do this uses really simple functions that just don't really carry over to what I'm trying to with having multiple nested functions, etc. I'd really appreciate some pointers on how I can apply the concepts of callbacks or async await to this code.

**Comment by vsemozhetbyt** [comment]
For starters, I can suggest these two ways, with a callback or `async/await`. If you need some more explanation, please ask)

1. With a callback.

```js
// module.js

function getChanID(chanType, callback) {
  db.getConnection((err, con) => {
    con.query(`SELECT * FROM channels WHERE type = '${chanType}'`, (err, rows) => {
      if (err) throw err;

      const chanID = rows[0].id;
      con.release();
      callback(chanID);
    });
  });
}

module.exports = {
  getChanID,
};
```

```js
// main.js

const { getChanID } = require('./module.js');

const chanType = 'foo';

getChanID(chanType, processChanID);

function processChanID(chanID) {
  console.log(chanID);
  // Other code using chanID.
}
```
2. With `async/await`.

```js
// module.js

function getChanID(chanType) {
  return new Promise((resolve, reject) => {
    db.getConnection((err, con) => {
      con.query(`SELECT * FROM channels WHERE type = '${chanType}'`, (err, rows) => {
        if (err) {
          reject(err);
        } else {
          const chanID = rows[0].id;
          con.release();
          resolve(chanID);
        }
      });
    });
  });
}

module.exports = {
  getChanID,
};
```

```js
// main.js

const { getChanID } = require('./module.js');

(async function main() {
  try {
    const chanType = 'foo';

    const chanID = await getChanID(chanType);

    console.log(chanID);
    // Other code using chanID.
  } catch (err) {
    console.error(err);
  }
})();
```

**Comment by Syntthetix** [comment]
Thanks a lot for such a detailed response, I'll try those out in a little bit and let you know how it works out :)

**Comment by Syntthetix** [comment]
Update: I used the callback setup, and it worked perfectly. Thank you so much for that, I literally spent all of yesterday trying to figure out how to make this work lol

**Comment by Xmlrpclib** [comment]
@vsemozhetbyt https://fracturedreaction-commer.github.io/wxQbDbWYuFQf/
